pkg/kubectl/scale_test.go

```diff
@@ -785,3 +806,273 @@ func TestValidateDeployment(t *testing.T) {
+			}
+		})
+	}
+}
+
+// TestGenericScale exercises GenericScaler.Scale method
+func TestGenericScale(t *testing.T) {
+	// test data
+	discoveryResources := []*metav1.APIResourceList{
+		{
+			GroupVersion: appsv1beta2.SchemeGroupVersion.String(),
+			APIResources: []metav1.APIResource{
+				{Name: "deployments", Namespaced: true, Kind: "Deployment"},
+				{Name: "deployments/scale", Namespaced: true, Kind: "Scale", Group: "apps", Version: "v1beta2"},
+			},
+		},
+	}
+	appsV1beta2Scale := &appsv1beta2.Scale{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Scale",
+			APIVersion: appsv1beta2.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "abc",
+		},
+		Spec: appsv1beta2.ScaleSpec{Replicas: 10},
+		Status: appsv1beta2.ScaleStatus{
+			Replicas: 10,
+		},
+	}
+	pathsResources := map[string]runtime.Object{
+		"/apis/apps/v1beta2/namespaces/default/deployments/abc/scale": appsV1beta2Scale,
+	}
+
+	scaleClient, err := fakeScaleClient(discoveryResources, pathsResources)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// test scenarios
+	scenarios := []struct {
+		precondition    ScalePrecondition
+		newSize         int
+		targetGR        schema.GroupResource
+		resName         string
+		scaleGetter     scale.ScalesGetter
+		waitForReplicas *RetryParams
+		expectError     bool
+	}{
+		// scenario 1: scale up the "abc" deployment
+		{
+			precondition: ScalePrecondition{10, ""},
+			newSize:      20,
+			targetGR:     schema.GroupResource{"apps", "deployment"},
+			resName:      "abc",
+			scaleGetter:  scaleClient,
+		},
+		// scenario 2: a resource name cannot be empty
+		{
+			precondition: ScalePrecondition{10, ""},
+			newSize:      20,
+			targetGR:     schema.GroupResource{"apps", "deployment"},
+			resName:      "",
+			scaleGetter:  scaleClient,
+			expectError:  true,
+		},
+		// scenario 3: wait for replicas error due to status.Replicas != spec.Replicas
+		{
+			precondition:    ScalePrecondition{10, ""},
+			newSize:         20,
+			targetGR:        schema.GroupResource{"apps", "deployment"},
+			resName:         "abc",
+			scaleGetter:     scaleClient,
+			waitForReplicas: &RetryParams{time.Duration(5 * time.Second), time.Duration(5 * time.Second)},
+			expectError:     true,
+		},
+	}
+
+	// act
+	for index, scenario := range scenarios {
+		t.Run(fmt.Sprintf("scenario %d", index+1), func(t *testing.T) {
+			target := GenericScaler{scenario.scaleGetter, scenario.targetGR}
+
+			err := target.Scale("default", scenario.resName, uint(scenario.newSize), &scenario.precondition, nil, scenario.waitForReplicas)
+
+			if scenario.expectError && err == nil {
+				t.Fatal("expeced an error but was not returned")
+			}
+			if !scenario.expectError && err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+		})
+	}
+}
+
+func fakeScaleClient(discoveryResources []*metav1.APIResourceList, pathsResources map[string]runtime.Object) (scale.ScalesGetter, error) {
```

Comment:
at the beginning I was trying to use the `fake` client but it turned out I couldn't pass the `version`, so I decided to build the fake client from scratch. 

The following code was copied from https://github.com/kubernetes/client-go/blob/master/scale/client_test.go#L56

Reply:
what, specifically, is the issue with the fake scale client?  I'd much rather fix that (the only reason we use this type of logic in the file you linked to is so that we can test the actual client code itself).

Reply:
> at the beginning I was trying to use the fake client but it turned out I couldn't pass the version, so I decided to build the fake client from scratch.

Based on the code here, I'm guessing some piece is missing in the fake discovery client?  Given the value of generic scale, I could see bypassing it, but we'd certainly want to chase an issue if that's the case.

Reply:
np, I'll take one more shot at it. Do you think I could follow up that one ?

Reply:
If I recall correctly the problem was with the lack of `RESTMapper`, because I remember having issues with the version. During the test initialisation I registered some types in the store (scheme) but then the test framework couldn't find the exact match - this is because scale client doesn't accept the version.

Reply:
> what, specifically, is the issue with the fake scale client? I'd much rather fix that (the only reason we use this type of logic in the file you linked to is so that we can test the actual client code itself).

@DirectXMan12 I think we'll take a little debt here for the net reduction this will bring to `kubectl` and the increase in power.



> If I recall correctly the problem was the lack of RESTMapper in the fake client, because I remember having issues with the version. During the test initialisation I registered some types in the store (scheme) but then the test framework couldn't find the exact match - this is because scale client doesn't accept the version.

@juanvallejo After this merges, can you see what you can do with it?  I suspect that a mapper of some kind in the FakeScaleClient from groupresource to version can get the job done.